### DISCUSSION TITLE
Modernizing for 2024+

### DISCUSSION THREAD

**Discussion by crcrewso** [qatrackplus/qatrackplus#667] [General]
I've started to explore updating the dependencies for QAtrack+ to use a currently supported Django LTS release and I'm hoping I can bring the requirements all the way up to 4.2 without needing to touch the database. Here's the [project plan](https://github.com/users/crcrewso/projects/6), it's a little rough right now. I'd be happy to take any suggestions or feedback. 

I think one thing that might be nice to do would be to reduce the installation/testing targets, if you could report your install configuration that would also be helpful.

**Comment by randlet** [comment]
Hey Cody,

"all the way up to 4.2 without needing to touch the database."  I'm not positive but there are likely at least a few db migrations in Django itself plus other third party packages.

"I think one thing that might be nice to do would be to reduce the installation/testing targets"
That would certainly be nice.  Supporting multiple platforms & db's was likely a mistake and I regret it almost every time I do a release!  Very rough estimate would be 90% of installs are on Windows with SQL Server (split across multiple Windows Server & Sql Server versions :/),  8% on Ubuntu with Postgres, 1% on Ubuntu with MySQL and maybe a sprinkling of other *nix OS's.  As a start, getting people who are using MySQL to migrate to Postgres would be a good start.

**Comment by bobred** [comment]
I have for the past few weeks been tinkering with a test version of QATrack with the aim of updating the version of Python and of Django. I am willing to share my findings/process with you and the group. I currently have a system running on Python 3.12.7 and Django 4.2.18. I have only now started testing this, initially to ensure all the pages display properly and then a more in-depth test of the functionality.

So far, all is well with the initial tests of the pages displaying correctly, except for the Create Service Event, it is missing the top portion of the page, as below.
![createsvc](https://github.com/user-attachments/assets/1eb27b38-e9ad-4618-9a28-bef9b00dbe1f)
It appears as though the service_event_detail.html is not being loaded, I am trying to track down where and if this is the case.

**Reply by crcrewso** [reply]
I'd love to take a look, are you familiar with how to fork a repository and work under your own account?

**Reply by bobred** [reply]
Hi

Sorry for the delay I was out of county today at a regional dosimetry audit meeting.

It took me a while, but my changes should be uploaded here https://github.com/bobred/qatrackplus-update/tree/master.



Disclaimer, I am no expert in Django, only a basic level of understanding, but I am wanting to learn.

The method I employed was first to upgrade python and any subsequent packages, then Django 4.2 and the subsequent packages. I manually updated the code in various files until I found the package Django-upgrade with



git ls-files -- '*.py' | %{django-upgrade --target-version 4.2 $_}



I was able to bring the code up to 4.2 standards. You will probably neet to perform a makemigrations and migrate. 



If you have any questions, please get in touch.



James

**Reply by crcrewso** [reply]
Thank you for putting it up. I've added you to a github project page I tried to get started for this work.

**Reply by bobred** [reply]
Hi, regarding the Create Service Event issue I seem to have found where it is failing, though I don't know why. 
I have been trying to debug it for a while now, I have two versions running, the published version and my version, Django 4.2. Looking at the context data in views.py class ServiceEventUpdateCreate, I can't see any obvious differences between the two. The issue occurs in the template service_event_template.html, line 80, with the code {% for field in form.fieldsets.service_status %}. In the published version the for loop executes but in version 4.2 the for loop does not. I can't see any differences in form.fieldsets.service_status between the two versions. Any suggestions?

**Reply by crcrewso** [reply]
@bobred I'm off for the next couple days, so I can take a look then. 
Could you please detail as much as you can at [this link](https://github.com/medphys-canada/qatrackplus/issues/new/choose)

You can provide links to the bits of the code by going to the file of interest on github, clicking the line number, hitting the three dots, and select copy permalink.

**Reply by crcrewso** [reply]
@bobred 
I've been trying to test your update but I'm having some issues. What python version did you use? Did you run the test suite?

**Comment by riblettmj** [comment]
Fantastic work!  Randy pointed me to this conversation to connect with you guys -- I'm very interested in helping maintain QATrack+ going forward.  This is partly self-serving (I've deployed QATrack+ across a few sites that I hope will run for many years to come) but mostly because I've long benefited from the open-source work of others and want to pay it forward. 

With that in mind, I'm pulling down the fork now and will start playing around.

**Comment by bobred** [comment]
Hi

Initially 3.12 but I have tried it with 3.13. I ran it using PyCharm and VS Code. I did start running the tests and mainly found minor issues with the date formats.

I will have to have another look at it again, been busy trying to sell my house.

 

James

**Reply by crcrewso** [reply]
Sounds good. I'll be spending some time on it this weekend. 
Good luck, house sales are not the most fun things to do

**Comment by bobred** [comment]
HiGoing forward, if there is anything I can do to help please let me know.
James Murphy
 

    On Tuesday, 25 March 2025 at 17:27:19 GMT, Cody Crewson ***@***.***> wrote:  
 
 


@bobred
I've been trying to test your update but I'm having some issues. What python version did you use? Did you run the test suite?

**Comment by bobred** [comment]
HiHope things are good with you, did youget a chance to look at what I have done so far?James

 

    On Friday, 28 March 2025 at 15:42:51 GMT, Cody Crewson ***@***.***> wrote:  
 
 


Sounds good. I'll be spending some time on it this weekend.
Good luck, house sales are not the most fun things to do

**Reply by crcrewso** [reply]
I did, current package management is quite an issue with trying to maintain and upgrade. I'm going to have to look at transitioning to poetry or an alternative. Trying to get the Administration page working properly was a challenge. Seeing what versions you chose certainly has helped but there are still issues unfortunately.

**Comment by bobred** [comment]
Hi

I have had another play with my GitHub branch and had to tweak a few bits, I’ll update the branch. I did forget about the issue with the Django-admin-views package, https://github.com/koleror/django-admin-views/pull/43/commits/dffbc56f4f6b271c59c7e8bdc2036e8cbd5c4589.

 

I have taken a snapshot of the working QATrack app and uploaded it to my Google drive, I can share the link, everything seems to be working apart from the create service event that I have mentioned previously.

 

James

**Reply by crcrewso** [reply]
That's great news. Don't worry too much about admin-views. @ETS1199 is working on removing/replacing that functionality with a more modern, currently maintained version. We're going to roll your changes together with his and mine for the next release. If you'd like the code to continue to show your contributions it would be helpful to keep your code updated on github. We can grab your commits as we validate them.

**Reply by bobred** [reply]
Thats great. Any thoughts on the create service event only parially rendering the page? Its been driving me mad.

**Reply by crcrewso** [reply]
Not quest yet, that's probably a next week thing. Right now we're actually looking at getting rid of both Django-Form-Utils and Django-Admin-Views entirely and replacing them with supported alternatives. That will require a rewrite to the parts of the create service event page that's giving you trouble.

**Comment by leducvin** [comment]
Happily running on docker in a debian container with MySQL and nginx with QAT v3.1.1.3. We made our own docker setup starting with the official python image, but I know there's also https://hub.docker.com/r/simonbiggs/qatrackplus.

Any plans on keeping docker as an official deployment method? It makes applications nice and easy to install, test and deploy usually.

Would certainly consider switching from MySQL to Postgres, but I'm not sure how, if any special steps are required. Maybe writing a migration guide would help users to do that.

**Reply by crcrewso** [reply]
The only thing standing in the way from properly supporting docker would be sufficient testing. If you're willing, could you take a look at the Dev-SCA branch and see if that version is still compatible with Docker. Of major note is the transition from requirements.txt to uv/python.toml standards.

**Reply by leducvin** [reply]
Sure, I'll give that branch a try under docker and let you know.

**Reply by leducvin** [reply]
What's under deploy/docker won't work anymore for the Dev-SCA branch. It's based on python 3.6 (and I see you're now targeting python 3.12 or 3.13). And it installs dependencies from an external repo:

```dockerfile
RUN wget https://bitbucket.org/SimonGBiggs/qatrackplus/raw/simon-docker/requirements.postgres.txt
RUN wget https://bitbucket.org/SimonGBiggs/qatrackplus/raw/simon-docker/requirements.txt
```

I think it would be better just to build the image from the official QAT repo. These are only the problems I can see at first glance, I'm sure there's more.

If there's interest, I don't mind working on a docker deploy method for the next release. I'll have to do this for our own local installation anyway. I don't see any major problem with the switch to pyproject.toml or uv.

**Reply by crcrewso** [reply]
We just noticed a few issues with the presence of a couple dependencies that are causing issues with other test installs. Once these have been sorted it would be great if you could look into a docker compose. For ReadItLater I'm having to use UV to create a requirements.txt file. A future question, would it be more appropriate to use uv or provide a requirements.txt file for Docker. I'm okay with either based on your recommendation.

**Reply by leducvin** [reply]
Sure thing. I think we'll be able to use uv. I'll try in any case. Let me know when you think you've got the requirements stable and dependencies figured out.

**Comment by leducvin** [comment]
Hello, I'm interested in volunteering to test upgrading our v3.1.1.4 version database to v4.0Beta.2 (or is it too early for that?).

At the same time, I'd like to figure out a docker deployment method.

Can I use the v4.0Beta.2 tag as a starting point for this work? Or rather Dev or Dev-SCA branch? Are the requirements and dependencies pretty stable at this point?

**Reply by crcrewso** [reply]
That would be perfect, and that's what that tag is for. For windows deployments I think I'm going to transition to Apache, and the setup instructions for that version for developers should get you started. Please start a new issue if you see anything of concern!

**Reply by leducvin** [reply]
Great! I think I'm going to start on this next week

**Comment by nsmela** [comment]
Hi everyone!

I'd love to help modernizing this repo. Got a docker setup to work with python 3.12 and a development docker setup. [This](https://github.com/nsmela/qatrackplusplus) branch also uses Github Actions to evaluate the code formatting, imports formatting, and testing. These changes have not been tested against the manual installation instructions, but planning to run them in virtual machines for that.

I wanted to touch base about the roadmap for the future and a list of priorities. I'm not an experienced in running a development project, but this is what I'd suggest:

1. **Limit Support Scope:** Completely freeze any features developing for now. As randlet noted, development time might be disproportionately spent on features of little use.  The team should list what setups will be within scope and supported going forward. Personally, I'd limit it to ngix, postgresql, and sql server. We'd need a way to allow existing users to upgrade from older unsupported database systems.
2. **Docker Files:** have a standard one for deployments and add a dev-focused one for evaluating changes or for testing
3. **Testing and Evaluating:** set up a process to ensure future submissions follow the project's standards and pass testing setups. Do our best to catch breaking changes at the PR level
4. **Installations, Upgrades, Backing up and Restoring:** I've been testing a boostrap script where a single line in powershell installs all the required programs and installs another script to install, backup, restore, and upgrade qatrackplus for the user.  The user doesn't even need the bootstrap script on their hard drive, the command copies the latest version of the scriptfrom a github repo, installs it, and then runs it.

**Reply by crcrewso** [reply]
Welcome to the team! The one feature that we have to take live quickly is the localization addition. I received the CNSC Innovation grant for that feature's deployment. The only holdback on the new Dev branch is windows deployment. If you want to write up an install guide for windows with IIS/CherryPy (I think that's what you said you were working with), go for it. I'll test it out. 

My one concern with Docker is that I'd like either focus, dev or prod, to be OS agnostic, so if you can install docker desktop, you can install QATrack+. Do you think that's something possible with what you've seen? 

The holdback on supporting database transitions is testing. I would be very open to supporting 6 sites this year, with remote support navigating their specific migration path, as long as the site would let me publish the workflow.

**Reply by nsmela** [reply]
> If you want to write up an install guide for windows with IIS/CherryPy (I think that's what you said you were working with), go for it. I'll test it out.
One of my first priorities. I've gotten it to work on [this repo](https://github.com/nsmela/qatrackplusplus/tree/master) and am workign to bring the lessons over. Switched from black, isort, etc to Ruff and created a slew of changes. Working on making the PR smaller and more managable.

> My one concern with Docker is that I'd like either focus, dev or prod, to be OS agnostic, so if you can install docker desktop, you can install QATrack+. Do you think that's something possible with what you've seen?
From my limited experience, docker always runs in Linux. If running docker on Windows, it creates a linux vm to run it. But I could be wrong. This [branch](https://github.com/nsmela/qatrackplusplus/tree/master) has a prod and dev docker.

> The holdback on supporting database transitions is testing. I would be very open to supporting 6 sites this year, with remote support navigating their specific migration path, as long as the site would let me publish the workflow.
I need an example database to test. Asked my leadership for a sql_dump of our test db and a zip file of our media, then I can try on a server we've actively used.

**Reply by crcrewso** [reply]
> I need an example database to test. Asked my leadership for a sql_dump of our test db and a zip file of our media, then I can try on a server we've actively used.

https://share.google/aimode/rw8YWypPMlx4PcTHb

I hope this helps, this AI response seems to cover all of the details I remember regarding exporting a Django DB using the native tools. It can then be imported into sqlite (or any other db) using the native importing tools. This is the mechanism I was thinking of using for supporting MySQL migrations. 

>From my limited experience, docker always runs in Linux.

That's not entirely true, there are ways to build a container using docker that runs on top of a Windows Server install, it's not common unless you're deploying very special cases, so I think it's quite safe to forget it exists. I found it because there was some docker container I tried to run on Linux that failed completely. 

> Switched from black, isort, etc to Ruff

Amazing, I love Ruff and UV. Merging the commits should be much easier if we're both using the same tools. I haven't spent much time writing appropriate Ruff configurations so I'll take a look at yours.

**Comment by leducvin** [comment]
Just to chime in on docker here (which I wanted to work on, but still haven't had the time to do so...) @nsmela, I'm happy to start from what you have.

I think it's a great deployment option that can work on linux or Windows. Docker desktop on Windows can run linux containers without issue, as long as you either use the Hyper-V backend or the Windows Subsystems for linux (WSL 2) backend. I guess WSL 2 is recommended nowadays.

I don't think we should impose docker deployment, but having it available to both Windows and linux users would certainly be nice. It should make the installation of the whole app much simpler.

For docker, there should probably be a development docker-compose.dev.yaml file and a production docker-compose.prod.yaml file that define the services to run in either case.

Some things we need to think about in a deployment of this type is how to make upgrades and backups of the DB. Also how the uploaded files will be stored, how PDF reports will be generated.

I'm still hoping to try to migrate our MySQL DB to postgresql on the Dev branch. I'll be happy to share by experience once I get around to it.

If you need any help on localization, we're also happy to help as we'd like one day to have our own locally deployed version localized in French.

**Reply by crcrewso** [reply]
There is the start of a French localization file [here](https://github.com/qatrackplus/qatrackplus/tree/Dev/qatrack/locale/fr/LC_MESSAGES) I would love it if you could review it and make comments on any poorly translated statement. This infrastructure is how Django likes to deal with languages going forward. I still need to figure out if Canadian vs Parisian French is a distinction this product needs to make, or if Canadian French is good enough for everyone.

**Reply by nsmela** [reply]
> Just to chime in on docker here (which I wanted to work on, but still haven't had the time to do so...) @nsmela, I'm happy to start from what you have.

Here's my progress so far:
https://github.com/nsmela/qatrackplusplus/tree/master/deploy/docker

Here's a test repo for the scripts I tried out:
https://github.com/nsmela/qatrackplus-manager

Or, if you can give me guidelines and specifics, I can focus on those. I've been experimenting and getting for a feel for it so far.

> Some things we need to think about in a deployment of this type is how to make upgrades and backups of the DB. Also how the uploaded files will be stored, how PDF reports will be generated.

Big priority for me, plan on using scripts. Zip the media files, sql_dump of the DB. No experience with the pdfs. 

> I'm still hoping to try to migrate our MySQL DB to postgresql on the Dev branch. I'll be happy to share by experience once I get around to it.

If I gave you a script, would you give it a try? We'd start with a backup and restore script. Backup your install, use the backup to start a second server we could test.

**Reply by leducvin** [reply]
> There is the start of a French localization file [here](https://github.com/qatrackplus/qatrackplus/tree/Dev/qatrack/locale/fr/LC_MESSAGES) I would love it if you could review it and make comments on any poorly translated statement. This infrastructure is how Django likes to deal with languages going forward. I still need to figure out if Canadian vs Parisian French is a distinction this product needs to make, or if Canadian French is good enough for everyone.

I think there's a way to have international French (a choice of words that we can all work with), if we don't want to over-complicate things. I'll take a look at the files.

**Reply by leducvin** [reply]
I took a look at the docker setup and I can see it relies heavily on python scripts for various tasks. I'm not sure how to use it all, maybe writing a simple docker deployment guide in `README.md` would be of help. There's only a link in there and it's dead.

***

Here's some specific things I noticed:

The docker-compose in there under [0.3.0.dev](https://github.com/nsmela/qatrackplusplus/tree/master/deploy/docker/hub/0.3.0.dev) uses the image simonbiggs/qatrack:0.3.0.dev. We'd need to change that to some official image on dockerhub for the project.

The Dockerfiles also don't use multistage builds. qatrack is installed with its dependencies but also build-time dependencies like gcc. Also uv is bundled in the image. Better to install qatrack, in a build stage and copy over its virtual environment in a final stage where also its run-time dependencies are included, but not uv or any build time dependencies.

LDAP seems to be missing from the qatrack image.

***

I was thinking of a similar setup where you have docker compose with these services:

## Services

### qatrack

- runs the django app with gunicorn
- configure the app with env variables
- image: official python image with qatrackplus installed and some runtime dependencies (e.g. LDAP, supervisor). We supply our own Dockerfile.

### chrome
- allows qatrack to render HTML reports to PDF using chrome
- avoids installing chrome in the qatrack image
- depends on qatrack service (mounts a shared volume seeded with the qatrackplus code by the qatrack service, needs to read qatrack's CSS to render reports)
- image: mcr.microsoft.com/playwright (or some other image that runs chrome headless)

### djangoq
- runs supervisord+django Q (for running background tasks such as generating reports, sending emails)
- depends on the chrome service and the qatrack service (mounts a shared volume seeded with the qatrackplus code by the qatrack service)
- same image as qatrack service, just runs as a separate container

### proxy

- reverse proxy that serves ports 80 and 443 SSL
- communicates with gunicorn in the qatrack container
- image: official nginx image

### pgsql
- postgres SQL DB
- image: official postgresql image

### dbadmin
- provides an interface to browse tables, run SQL, etc.
- image: [adminer](https://hub.docker.com/_/adminer/)

## Volumes
- one for the database (used by postgres container)
- one for static files (used by qatrack and nginx containers)
- one for media files (i.e. uploaded files for upload tests in qatrack, used by qatrack and nginx containers)

***

For tooling we could use something like [just](https://just.systems/). It's a command runner that supports linux and Windows.

Basically, you define recipes. For example:

justfile
```
up:
  docker compose up
```
and then call it from the command-line as `just up`.

It can source .env files, so you can have you secrets configured in there. This allows us to turn hard to remember command lines into easy to remember:

justfile
```
set dotenv-load := true          # loads .env file in current directory

take-backup:
  docker compose exec pgsql pg_dump -U ${PG_USER} -F c ${PG_DB_NAME} > backup.dump
```

with .env:

```
PG_USER=myuser
PG_DB_NAME=mydbname
# other variables...
```

The variables in .env can be used in the compose file and passed on to the various services, including qatrack's django app. So we could also use those to configure the app.

***

The layout would look something like:
```
qatrackplus/
├── compose.yaml
├── compose.dev.yaml
├── .env.example <-- just shows some examples. The user is responsible for writing its own .env file
├── .dockerignore
├── justfile
├── docker/
│   ├── app/
│   │   ├── Dockerfile
│   │   ├── entrypoint.sh
│   │   └── gunicorn.conf.py
│   └── nginx/
│       └── default.conf.template
├── requirements/
│   ├── base.txt
│   ├── dev.txt
│   └── prod.txt
├── manage.py
├── qatrack/
├── docs/
└── ...
```

***

These are just ideas, patterns I've noticed in some other open-source projects. Like the tendency to configure everything with env variables, and use official docker images when possible. Let me know what you think!

Disclaimer: I used a LLM to provide me with an idea of best-practices docker deployment for a django app. It might be an echo chamber to some of my initial ideas...

**Reply by nsmela** [reply]
which branch are you on? I thought 0.3.x is cleared out

**Reply by leducvin** [reply]
Hello again, sorry I've not contributed more recently, been on vacation and busy in the clinic. On your repo, I was looking at the [0.3.0.dev](https://github.com/nsmela/qatrackplusplus/tree/master/deploy/docker/hub/0.3.0.dev) branch, looks like it's now gone.

For this project I'm using the "Dev" branch. I guess that's now "develop" I should be using.

**Reply by crcrewso** [reply]
Dev/develop are basically the same right now. I'm not sure how we'll manage Dev vs master vs develop in the near future.

For dev work, I'd recommend branching off of develop for now. 

Releases/4.0 would be the version we're intending to support for production installs, with other release branches for historical access. Tags work as well, including old tags. I'll cherry-pick commits into 4.0 for bug fixes and documentation as makes sense, and the rest will roll into Releases/4.1 when we decide to start beta testing that. 

I recently got the historical releases to build documentation with current ReadTheDocs configurations too.

**Comment by leducvin** [comment]
> If I gave you a script, would you give it a try? We'd start with a backup and restore script. Backup your install, use the backup to start a second server we could test.

Sure, thanks. What does the script do? It converts MySQL instructions from a MySQL backup file to postgresql instructions?

**Reply by nsmela** [reply]
it'd automate the steps in powershell for you

**Reply by leducvin** [reply]
I'm on linux, I don't know how much I can run powershell. I guess it's possible now. I'm still interested in trying to migrate from QATrack+ v3 on MySQL to v4 on postgresql.

**Reply by crcrewso** [reply]
For now I'd recommend v4 on MySQL. I'm actively working on migrating db from one type of server to another, and that should be version independent. You shouldn't need powershell on the Linux install. It would probably be easiest to install the front end on a fresh VM server. RedHat based (9+) and Ubuntu 22.04 LTS or newer (LTS) should both work well. 

Let us know where you get stuck so we can expand the documentation appropriately.

**Reply by leducvin** [reply]
OK, thanks. Maybe I'll leave it in MySQL to be safe. Do you know if I can swap MySQL for mariadb transparently? Most linux distributions only include mariadb-client now. I'll want to use the mariadb docker image for running the DB server, probably.

**Reply by crcrewso** [reply]
That's a question for someone who knows Django better, we're not doing anything to limit which SQL server you want to use but the project really needs to cut down the possible supported configurations until we have more admins. I've only used Microsoft SSMS and DBeaver for GUI SQL management, DBeaver does support multiple servers, and is available via most repositories, and flatpak. If you get a different DB working, please share how, and we'll absolutely find a way to keep the information public.
